### PR TITLE
feat: add feature flags ui

### PR DESCRIPTION
**Pull request by JivusAyrus**
## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: controlplane/src/core/repositories/FeatureFlagRepository.ts [hunk 1]
```diff
@@ -142,6 +154,193 @@ export class FeatureFlagRepository {
...
+      // Left join because version is optional
+      .leftJoin(schemaVersion, eq(schema.subgraphs.schemaVersionId, schema.schemaVersion.id))
+      .orderBy(asc(targets.createdAt), asc(schemaVersion.createdAt))
+      .where(and(...conditions))
+      .limit(limit)
+      .offset(offset);
+
+    const featureSubgraphs: SubgraphDTO[] = [];
+
+    for (const f of featureSubgraphTargets) {
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FeatureFlagRepository.ts:249]: FYI you can `for await` and iterate over the list directly.

File: controlplane/src/core/repositories/FeatureFlagRepository.ts [hunk 2]
```diff
@@ -403,6 +667,7 @@ export class FeatureFlagRepository {
         isFeatureGraph: subgraphs.isFeatureSubgraph,
         baseSubgraphId: featureSubgraphsToBaseSubgraphs.baseSubgraphId,
         isEventDrivenGraph: subgraphs.isEventDrivenGraph,
+        isFeatureSubgraph: subgraphs.isFeatureSubgraph,
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FeatureFlagRepository.ts:670]: Should we go with `isFeatureSubgraph` or `isFeatureGraph` ?

File: controlplane/src/core/repositories/FeatureFlagRepository.ts [hunk 3]
```diff
@@ -451,43 +716,46 @@ export class FeatureFlagRepository {
     return featureGraphsByFlag;
   }
 
-  // evaluates all the ffs which have fgs whose base subgraph id is the passed as input and returns the ffs that should be composed
-  public async getEnabledFeatureFlagsBySubgraphIdAndLabels({
-    subgraphId,
+  // evaluates all the ffs which have fgs whose base subgraph id and fed graph label matchers are passed as input and returns the ffs that should be composed
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FeatureFlagRepository.ts:719]: Please write it out. What means ffs?

File: controlplane/src/core/repositories/FeatureFlagRepository.ts [hunk 4]
```diff
@@ -580,11 +851,153 @@ export class FeatureFlagRepository {
...
+        routerConfigSignature: composition.routerConfigSignature || undefined,
+        admissionError: composition.admissionError || undefined,
+        deploymentError: composition.deploymentError || undefined,
+        isComposable: composition.isComposable || false,
+      });
+    }
+    return featureFlagCompositions;
+  }
+
+  public async getFeatureFlagSchemaVersionsByBaseSchemaVersion({
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FeatureFlagRepository.ts:915]: Please add comments to every method.

File: controlplane/src/core/repositories/FeatureFlagRepository.ts [hunk 5]
```diff
@@ -580,11 +851,153 @@ export class FeatureFlagRepository {
...
+  }
+
+  public async getFeatureSubgraphsByFeatureFlag({
+    featureFlagId,
+    namespaceId,
+  }: {
+    namespaceId: string;
+    featureFlagId: string;
+  }) {
+    const featureSubgraphsByFf = await this.db
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FeatureFlagRepository.ts:973]: Do we need to filter for the orgId here?

File: controlplane/src/core/repositories/FederatedGraphRepository.ts [hunk 6]
```diff
@@ -888,6 +891,42 @@ export class FederatedGraphRepository {
...
+          eq(targets.organizationId, this.organizationId),
+          eq(schemaVersion.id, data.schemaVersionId),
+          and(
+            eq(graphCompositions.isComposable, true),
+            or(isNull(graphCompositions.deploymentError), eq(graphCompositions.deploymentError, '')),
+            or(isNull(graphCompositions.admissionError), eq(graphCompositions.admissionError, '')),
+          ),
+        ),
+      )
+      .orderBy(desc(graphCompositions.createdAt))
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FederatedGraphRepository.ts:915]: Copy / paste issue?

File: studio/src/components/feature-flags-table.tsx [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+              command: `npx wgc feature-subgraph create <feature-subgraph-name> --namespace ${router.query.namespace} -r <routing-url> --subgraph <base-subgraph-name>`,
+            },
+            {
+              description:
+                "Publish a feature subgraphs using the below command.",
+              command: `npx wgc subgraph publish <feature-subgraph-name> --namespace ${router.query.namespace} --schema <schema-path> `,
+            },
+            {
+              description: "Create a feature flag using the below command.",
+              command: `npx wgc feature-flag create <feature-flag-name> --namespace ${router.query.namespace} --label ${label} --feature-subgraphs <feature-subgraph-names...> -e`,
```
Comment (StarpTech) [inline studio/src/components/feature-flags-table.tsx:68]: The fourth step is to enable the feature flag.

File: studio/src/components/layout/feature-flag-layout.tsx [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  const links: NavLink[] = useMemo(() => {
+    const basePath = `/${organizationSlug}/${namespace}/feature-flag/${slug}`;
+
+    return [
+      {
+        title: "Feature subgraphs",
+        href: basePath + "/feature-subgraphs",
+        icon: <Component2Icon className="h-4 w-4" />,
+      },
+      // {
```
Comment (StarpTech) [inline studio/src/components/layout/feature-flag-layout.tsx:72]: Remove?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: controlplane/src/core/repositories/FeatureFlagRepository.ts [hunk 9]
```diff
@@ -142,6 +154,193 @@ export class FeatureFlagRepository {
...
+      .where(and(...conditions))
+      .limit(limit)
+      .offset(offset)
+      .execute();
+
+    return resp.map((r) => ({
+      ...r,
+      labels: r.labels?.map?.((l) => splitLabel(l)) ?? [],
+      createdAt: r.createdAt.toISOString(),
+      updatedAt: resp[0].updatedAt?.toISOString() || '',
```
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply controlplane/src/core/repositories/FeatureFlagRepository.ts:191]: Done.

File: controlplane/src/core/repositories/FeatureFlagRepository.ts (hunk 2 above)
Comment (JivusAyrus) [reply controlplane/src/core/repositories/FeatureFlagRepository.ts:670]: We call it Feature Subgraph, so isFeatureSubgraph

File: controlplane/src/core/repositories/FeatureFlagRepository.ts (hunk 3 above)
Comment (JivusAyrus) [reply controlplane/src/core/repositories/FeatureFlagRepository.ts:719]: Done

File: controlplane/src/core/repositories/FeatureFlagRepository.ts (hunk 5 above)
Comment (JivusAyrus) [reply controlplane/src/core/repositories/FeatureFlagRepository.ts:973]: Done

File: controlplane/src/core/repositories/FederatedGraphRepository.ts (hunk 6 above)
Comment (JivusAyrus) [reply controlplane/src/core/repositories/FederatedGraphRepository.ts:915]: Yep, Done

File: studio/src/components/feature-flags-table.tsx (hunk 7 above)
Comment (JivusAyrus) [reply studio/src/components/feature-flags-table.tsx:68]: We enable the feature flag while creating it itself

File: studio/src/components/layout/feature-flag-layout.tsx (hunk 8 above)
Comment (JivusAyrus) [reply studio/src/components/layout/feature-flag-layout.tsx:72]: Done

File: studio/src/components/feature-flags-table.tsx (hunk 7 above)
Comment (StarpTech) [reply studio/src/components/feature-flags-table.tsx:68]: Can we use the full flag name? I think `--enabled` is easier to understand for beginners.

Comment (JivusAyrus) [reply studio/src/components/feature-flags-table.tsx]: Done

**Review by StarpTech — APPROVED** [review]
LGTM

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 10]
```diff
@@ -8146,6 +8223,41 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+          if (!featureFlag) {
+            return {
+              response: {
+                code: EnumStatusCode.ERR_NOT_FOUND,
+                details: `Feature flag ${req.featureFlagName} not found`,
+              },
+            };
+          }
+
+          const ffSchemaVersion = await featureFlagRepo.getFeatureFlagSchemaVersionByBaseSchemaVersionAndFfId({
```
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply controlplane/src/core/bufservices/PlatformService.ts:8240]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 11]
```diff
@@ -8247,6 +8360,9 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
         if (!federatedGraph) {
           return {
             subgraphs: [],
+            featureFlags: [],
+            featureFlagsInLatestValidComposition: [],
```
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply controlplane/src/core/bufservices/PlatformService.ts:8364]: Cant think of anything meaningful

[the remark this answers is not in the document]
Comment (StarpTech) [reply controlplane/src/core/bufservices/PlatformService.ts:8364]: ok

**Pull request merged**